module/zfs/spa.c

```diff
@@ -2316,6 +2320,166 @@ vdev_count_verify_zaps(vdev_t *vd)
 }
 #endif
 
+static unsigned long
+spa_get_hostid(void)
+{
+	unsigned long myhostid;
+
+#ifdef	_KERNEL
+	myhostid = zone_get_hostid(NULL);
+#else	/* _KERNEL */
+	/*
+	 * We're emulating the system's hostid in userland, so
+	 * we can't use zone_get_hostid().
+	 */
+	(void) ddi_strtoul(hw_serial, NULL, 10, &myhostid);
+#endif	/* _KERNEL */
+
+	return (myhostid);
+}
+
+/*
+ * Determine whether the activity check is required.
+ */
+static boolean_t
+spa_activity_check_required(nvlist_t *config, uberblock_t *ub, spa_t *spa)
+{
+	uint64_t config_pool_state = 0;
+	uint64_t hostid = 0, myhostid = 0;
+	uint64_t orig_txg = 0;
+	uint64_t orig_timestamp = 0;
+
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_IMPORT_TXG,
+	    &orig_txg);
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_TIMESTAMP,
+	    &orig_timestamp);
+	dprintf("mmp spa_activity_check_required orig_txg %llu "
+	    "orig_timestamp %llu\n", (u_longlong_t)orig_txg,
+	    (u_longlong_t)orig_timestamp);
+
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_STATE,
+	    &config_pool_state);
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_HOSTID, &hostid);
+	myhostid = spa_get_hostid();
+
+	dprintf("mmp spa_load_impl pool_state %llu hostid %llu myhostid %llu\n",
+	    config_pool_state,
+	    (u_longlong_t)hostid,
+	    (u_longlong_t)myhostid);
+
+	/*
+	 * Special values to indicate test should be skipped, for zdb
+	 */
+	if (orig_txg == MMP_SKIP_ACTIVITY_TEST ||
+	    spa->spa_skip_activity_chk == B_TRUE)
+		return (B_FALSE);
+
+	/*
+	 * Since the hostid is likely to be different before the pivot than
+	 * it is after the pivot, and since it is hard to see how two nodes
+	 * could safely boot from the same pool anyway, we do not do MMP
+	 * for root pools.
+	 */
+	if (spa_is_root(spa))
+		return (B_FALSE);
+
+	/*
+	 * Skip the test if the current host, and the host who last imported the
+	 * pool, both have (or had) MMP disabled.  If the pool is successfully
+	 * imported, we will then respect the local setting for mmp writes.
+	 */
+	if (ub->ub_mmp_magic == MMP_MAGIC && ub->ub_mmp_delay == 0 &&
+	    zfs_mmp_interval == 0)
+		return (B_FALSE);
+	/*
+	 * If the orig_* values are nonzero, they are the results of an earlier
+	 * tryimport.  If they match the uberblock we just found, then the pool
+	 * has not changed and we return false so we do not test a second time.
+	 */
+	if (orig_txg && orig_timestamp && orig_txg == ub->ub_txg &&
+	    orig_timestamp == ub->ub_timestamp)
+		return (B_FALSE);
+
+	if (hostid == myhostid)
+		return (B_FALSE);
+
+	if (config_pool_state != POOL_STATE_ACTIVE)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+static int
+spa_activity_check(uberblock_t *ub, spa_t *spa)
+{
+	kcondvar_t cv;
+	kmutex_t mtx;
+	hrtime_t import_intervals, activity_test_duration;
```

Comment:
It seems `activity_test_duration` is used as an absolute time, not a duration. Perhaps it would be better named `activity_test_deadline` or `activity_test_end_time`.